File: chapitre7/affiche.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

// Fonction exécutée lors de la réception du signal [8, 9]
void gestionnaire(int numero) {
    if (numero == SIGUSR1) { // Vérification du type de signal [7]
        FILE *fp = fopen("/tmp/entier.txt", "r");
        if (fp != NULL) {
            int n;
            fscanf(fp, "%d", &n); // Lecture de l'entier dans le fichier [5]
            printf("\n[NOTIFICATION] L'entier enregistré est : %d\n", n);
            fclose(fp);
            exit(0);
        } else {
            printf("\nErreur : Impossible de lire /tmp/entier.txt\n");
        }
    }
}

int main(void) {
    // Affichage du PID pour permettre la communication avec saisit.c [10]
    printf("Mon PID est : %d\n", getpid());

    struct sigaction action;
    action.sa_handler = gestionnaire; // Association de la fonction de capture [9]
    sigemptyset(&action.sa_mask);
    action.sa_flags = 0;

    // Enregistrement de la capture du signal utilisateur SIGUSR1 [9, 11]
    if (sigaction(SIGUSR1, &action, NULL) != 0) {
        perror("Erreur de configuration du signal");
        exit(1);
    }

    // Boucle d'attente infinie utilisant sleep pour économiser les ressources [1, 12]
    while (1) {
        printf("J'attends que saisit.c m'envoie un signal...\n");
        sleep(10); 
    }

    return 0;
}

File: chapitre7/affiche.py
import os
import signal
import time
import sys

# Fonction exécutée lors de la réception du signal (Signal Handler)
def gestionnaire(numero, frame):
    if numero == signal.SIGUSR1:  # Vérification du type de signal [4]
        try:
            # Ouverture et lecture de l'entier dans le fichier [3]
            with open("/tmp/entier.txt", "r") as fp:
                n = fp.read().strip()
                print(f"\n[NOTIFICATION] L'entier enregistré est : {n}")
                sys.exit(0)  # Terminaison propre après lecture
        except FileNotFoundError:
            print("\nErreur : Impossible de lire /tmp/entier.txt")

def main():
    # Affichage du PID pour permettre la communication avec saisit.py [1, 5]
    print(f"Mon PID est : {os.getpid()}")

    # Association de la fonction de capture pour SIGUSR1 [3]
    # Remplace l'utilisation de la structure sigaction en C
    signal.signal(signal.SIGUSR1, gestionnaire)

    # Boucle d'attente utilisant sleep pour économiser les ressources [3, 6]
    try:
        while True:
            print("J'attends que saisit.py m'envoie un signal...")
            time.sleep(10) 
    except KeyboardInterrupt:
        # Permet de quitter proprement avec Ctrl-C si besoin
        pass

if __name__ == "__main__":
    main()